In [99]:
import pandas as pd
import sklearn


In [100]:
train_df = pd.read_csv('C:/Users/Mirela/PycharmProjects/App/spam/train.csv').fillna(' ')
valid_df = pd.read_csv('C:/Users/Mirela/PycharmProjects/App/spam/valid.csv').fillna(' ')
test_df = pd.read_csv('C:/Users/Mirela/PycharmProjects/App/spam/test.csv').fillna(' ')


In [101]:
train_df

text  label
0              i m in solihull  | do you want anything       1
1                          gud mrng dear hav a nice day      1
2                   dude how do you like the buff wind       1
3     yup  wun believe wat  u really neva c e msg i ...      1
4     did you get any gift  this year i didnt get an...      1
...                                                 ...    ...
4731  hi  wlcome back  did wonder if you got eaten b...      1
4732                             sorry  i ll call later      1
4733      prabha  i m soryda  realy  frm heart i m sory      1
4734                         nt joking seriously i told      1
4735            did he just say somebody is named tampa      1

[4736 rows x 2 columns]

In [102]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4736 entries, 0 to 4735
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    4736 non-null   object
 1   label   4736 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 74.1+ KB


In [103]:
valid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 585 entries, 0 to 584
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    585 non-null    object
 1   label   585 non-null    int64 
dtypes: int64(1), object(1)
memory usage: 9.3+ KB


In [104]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    9 non-null      object
 1   label   9 non-null      int64 
dtypes: int64(1), object(1)
memory usage: 272.0+ bytes


In [105]:
train_text = train_df['text']
valid_text = valid_df['text']
test_text = test_df['text']
all_text = pd.concat([train_text,valid_text,test_text])

In [106]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(ngram_range=(1, 2),min_df=2,max_features=12000)

vectorizer = vectorizer.fit(all_text)
train_features = vectorizer.transform(train_text)
valid_features = vectorizer.transform(valid_text)
test_features = vectorizer.transform(test_text)

y_train = train_df['label']
y_valid = valid_df['label']
y_test = test_df['label']

print(y_train.shape)
print(y_valid.shape)
print(y_test.shape)
print(train_features.shape)
print(valid_features.shape)
print(test_features.shape)
    

(4736,)
(585,)
(9,)
(4736, 12000)
(585, 12000)
(9, 12000)


In [135]:
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import MultinomialNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import LinearSVC


In [136]:
lr = LogisticRegression(random_state=0).fit(train_features, y_train)
y_pred_clf=lr.predict(valid_features)

In [137]:
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix
"Accuracy score is {}".format(accuracy_score(y_valid,y_pred_clf))

'Accuracy score is 0.9623931623931624'

In [138]:
mn = MultinomialNB()
mn.fit(train_features, y_train)
y_pred_mn=mn.predict(valid_features)
"Accuracy score is {}".format(accuracy_score(y_valid,y_pred_clf))

'Accuracy score is 0.9623931623931624'

In [139]:
dt = DecisionTreeClassifier(random_state=0).fit(train_features, y_train)
y_pred_clf=dt.predict(valid_features)
"Accuracy score is {}".format(accuracy_score(y_valid,y_pred_clf))



'Accuracy score is 0.9641025641025641'

In [140]:
svc = LinearSVC(random_state=0).fit(train_features, y_train)
y_pred_clf=svc.predict(valid_features)
"Accuracy score is {}".format(accuracy_score(y_valid,y_pred_clf))

'Accuracy score is 0.9863247863247864'

Best classifier  is  : mn - MultinomialNB()

In [141]:
from sklearn.metrics import precision_score
"Precision score is {}".format(precision_score(y_valid,y_pred_mn))


'Precision score is 0.975517890772128'

In [142]:
from sklearn.metrics import recall_score
"Recall score is {}".format(recall_score(y_valid,y_pred_mn))

'Recall score is 1.0'

In [143]:
from sklearn.metrics import f1_score
"f1 score is {}".format(f1_score(y_valid,y_pred_mn))

'f1 score is 0.9876072449952336'

In [149]:
confusion_matrix(mn.predict(valid_features), y_valid)

array([[ 54,   0],
       [ 13, 518]], dtype=int64)

true positives (TP) = 518: These are cases in which we predicted yes (they have the spam), and they do have the spam.

true negatives (TN) = 54 : We predicted no, and they don't have the spam. 

false positives (FP) = 0: We predicted yes, but they don't actually have the spam. (Also known as a "Type I error.") 

false negatives (FN) = 13: We predicted no, but they actually do have the spam. (Also known as a "Type II error.")


In [150]:
mn.predict_proba(valid_features)[:,1]

array([0.99811424, 0.99158126, 0.93798258, 0.97991807, 0.98660516,
       0.3437895 , 0.93497377, 0.99046041, 0.98272141, 0.9877606 ,
       0.98367702, 0.99732368, 0.99788503, 0.99504164, 0.99257903,
       0.99846553, 0.02166372, 0.98045021, 0.90952619, 0.99355996,
       0.00101944, 0.99537018, 0.94847102, 0.9749498 , 0.97006802,
       0.98440352, 0.94715742, 0.01006353, 0.86216391, 0.97066818,
       0.98852855, 0.97417321, 0.98970752, 0.99919086, 0.99469794,
       0.98842492, 0.98693716, 0.95442558, 0.97527199, 0.9998968 ,
       0.98443318, 0.98500271, 0.00424268, 0.98656009, 0.98754059,
       0.25535414, 0.99291608, 0.99962449, 0.99831772, 0.99262517,
       0.92125855, 0.99735784, 0.97620124, 0.99406973, 0.99500968,
       0.96908463, 0.01806679, 0.96343802, 0.9481767 , 0.99521255,
       0.03042059, 0.99172681, 0.9982861 , 0.98068345, 0.13814138,
       0.97414555, 0.96785272, 0.98184815, 0.98364316, 0.96140106,
       0.98449819, 0.99640313, 0.99507404, 0.99909225, 0.67667

In [151]:
predictions = mn.predict_proba(test_features)[:, 1]
"Predictions of our classifier are:{}".format(predictions)

'Predictions of our classifier are:[0.99158126 0.06157155 0.99046041 0.5464834  0.9877606  0.98660516\n 0.3437895  0.42502005 0.98272141]'

Conclusions: 
Best classifier here is multinomialNB. I decided to calculate some metrics like precision_score, recall_score, f1_score. At the and I made some predictions for the test set. For this example, I used TfidfVectorizer to convert a collection of raw documents into a matrix of TF-IDF features. I decided to use the next classifiers : LogisticRegression, MultinomialNB,DecisionTreeClassifier, LinearSVC. 
